txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,482 @@
+            let i_amount = amount;
+            let i_color = color;
+            match output {
+                Output::Coin { to, amount, color } => {
+                    if to != i_owner {
+                        return Err(Error::NotInsertedIOWrongOwner.into());
+                    }
+                    if amount < i_amount {
+                        return Err(Error::NotInsertedIOWrongAmount.into());
+                    }
+                    if color != i_color {
+                        return Err(Error::NotInsertedIOWrongAssetIt.into());
+                    }
+                }
+                Output::Contract { .. } => return Err(Error::NotInsertedIOConractOutput.into()),
+                Output::Withdrawal { .. } => {
+                    return Err(Error::NotInsertedIOWithdrawalInput.into());
+                }
+                Output::Change { to, color, amount } => {
+                    if to != i_owner {
+                        return Err(Error::NotInsertedIOWrongOwner.into());
+                    }
+                    if color != i_color {
+                        return Err(Error::NotInsertedIOWrongAssetIt.into());
+                    }
+                    if is_output_filled && amount < i_amount {
+                        return Err(Error::NotInsertedIOWrongAmount.into());
+                    }
+                }
+                Output::Variable { to, amount, color } => {
+                    if is_output_filled {
+                        if to != i_owner {
+                            return Err(Error::NotInsertedIOWrongOwner.into());
+                        }
+                        if amount < i_amount {
+                            return Err(Error::NotInsertedIOWrongAmount.into());
+                        }
+                        if color != i_color {
+                            return Err(Error::NotInsertedIOWrongAssetIt.into());
+                        }
+                    }
+                    // else do nothing, everything is variable and can be only check on execution
+                }
+                Output::ContractCreated { .. } => {
+                    return Err(Error::NotInsertedIOConractOutput.into())
+                }
+            };
+        } else {
+            panic!("Use it only for coin output check");
+        }
+        Ok(())
+    }
+
+    /// insert tx inside dependency
+    /// return list of transactions that are removed from txpool
+    pub async fn insert<'a>(
+        &'a mut self,
+        txs: &'a HashMap<TxId, ArcTx>,
+        db: &dyn TxPoolDB,
+        tx: &'a ArcTx,
+    ) -> anyhow::Result<Vec<ArcTx>> {
+        let mut collided: Vec<TxId> = Vec::new();
+        // iterate over all inputs and check for colision
+        let mut max_depth = 0;
+        let mut db_coins: HashMap<UtxoId, CoinState> = HashMap::new();
+        let mut db_contracts: HashMap<ContractId, ContractState> = HashMap::new();
+        for input in tx.inputs() {
```

Comment:
This function has a lot of deeply nested complexity, while that may be unavoidable in situations with lots of inherent complexity, I think there's probably some room to refactor this into smaller units.

Reply:
Will consider it.